transforms/universal/aggregator/ray/src/aggregator_transform_ray.py

```diff
@@ -0,0 +1,339 @@
+        defined https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/put_object.html
+        in the case of failure dict is None and number of operation retries.
+        Retries are performed on operation failures and are typically due to the resource overload.
+        """
+        if len(self.words) == 0:
+            # No data
+            return {}, 0
+        output_folder = self.data_access.get_output_folder()
+        if not output_folder.endswith("/"):
+            output_folder += "/"
+        """
+        output_path = f"{output_folder}{self.id}.parquet"
+        # convert to parquet
+        words = pa.array(self.words.keys())
+        counts = pa.array(self.words.values())
+        names = ["words", "counts"]
+        table = pa.Table.from_arrays(arrays=[words, counts], names=names)
+        # save table
+        return self.data_access.save_table(path=output_path, table=table)
+        """
+        # convert dictionary to CSV
+        s = io.StringIO()
+        dict_writer = csv.DictWriter(s, ["word", "count"], extrasaction="ignore")
+        for i, datum in enumerate(self.words.items()):
+            row = {'word': datum[0], 'count': datum[1]}
+            dict_writer.writerow(row)
+        # reset cursor to the beginning of the StringIO stream
+        s.seek(0)
+        output_path = f"{output_folder}{self.id}.csv"
+        return self.data_access.save_file(path=output_path, data=s.read().encode('utf-8'))
+
+
+class AbstractTokenizer:
+    """
+    Abstract tokenizer class
+    """
+    def tokenize(self, text: str) -> list[str]:
+        """
+        Tokenize string
+        :param text: source text
+        :return: list of tokens (words)
+        """
+        pass
+
+
+class SimpleTokenizer(AbstractTokenizer):
+    """
+    Simple implementation of the abstract tokenizer
+    """
+    def tokenize(self, text: str) -> list[str]:
+        """
+        Tokenize string
+        :param text: source text
+        :return: list of tokens (words)
+        """
+        # start from normalizing string
+        normal = text.strip().lower().translate(str.maketrans("", "", string.punctuation))
+        return normal.split()
+
+
+class AggregateTransform(AbstractTableTransform):
+    """
+    Implements Aggregator table transformer.
+    """
+
+    def __init__(self, config: dict):
+        """
+        Initialize based on the dictionary of configuration information.
+        The dictionary should contain the following:
+            doc_column - name of the doc column
+            tokenizer - tokenizer to use
+            aggregators - list of aggregator actors, references
+        """
+        # Make sure that the param name corresponds to the name used in apply_input_params method
+        # of AggregateTableTransformConfiguration class
+        super().__init__(config)
+        self.doc_column = config.get("doc_column", "")
+        self.aggregators = config.get("aggregators", [])
```

Comment:
So, you're willing to make the tokenizer pluggable with AbstractTokenizer.  Why not make the aggregator pluggable with an AbstractAggregator?  Then this can become and python-only class. So, it can be done, it seems its just that you don't want to do it?

Reply:
Yes, there is a big difference between pluggable Python code - easy and pluggable actor or Python code - access API change

Reply:
I had done this for ededup and I believe it worked...

https://github.ibm.com/ai-models-data/fm-data-engineering/blob/ededup-conventions/transforms/universal/ededup/src/ededup_transform.py

Reply:
Sorry, please. If we decide we need python only version, then we will do it. We discussed it today